### PR TITLE
[expo-notifications] Do not backup installation ID

### LINKED ISSUE #2824: Push Notification token doesn't update with a new device
### Environment

Production Standalone iOS

### Steps to Reproduce

1. User installs the app
2. Enables push notifications
3. Get token via `Notifications.getExpoPushTokenAsync()`
4. User receives a push notification successfully
5. User makes a backup of their device
6. User gets a new device
7. User restores from backup
8. User opens already-installed app
9. Push token via `Notifications.getExpoPushTokenAsync()` is the same as in 3
10. No push notifications are ever received by the new device

### Expected Behavior

A new token should probably be generated so that the server can send to the new device.

### Actual Behavior

No push notifications are received by the new device.

Only course of action for the user seems to be to uninstall + reinstall, at which point, push notifications work without issue.

### LINKED ISSUE #6051: Push Notifications don't work on new phone that is restored from an old phone due to duplicate device ID
### Environment

(Relates to Expo Push Notification service)

### Steps to Reproduce

1) Set up a new iPhone, and choose to restore from a backup of an existing iPhone with the Expo client or production expo app.
2) Try sending a push notification using the Expo notification tool to the push token reported by the new phone (it'll be the same as the one on the old phone!)

### Expected Behavior

I would expect _both_ devices (old and new) to receive the notification, or at the very least the newest-registered device. 

### Actual Behavior

Only the old device receives the notification.

This is happening because both devices now have the *same* Device ID/push token, which is supposed to be unique across app installations. I guess restoring a phone from a backup doesn't count as a new installation in practice.

While ideally each device would have a unique ID, even if restored from the same backup, another solution would be to ensure all devices with the same ID receive push notifications.

### Reproducible Demo

I can't really do this since it requires restoring a phone from a backup :D

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

Fixes https://github.com/expo/expo/issues/6051. Fixes https://github.com/expo/expo/issues/2824.

# How

On iOS I moved the installation ID from `NSUserDefaults` which are always synced and included in backup to a simple `.txt` file in the documents directory to which I add a [`NSURLIsExcludedFromBackupKey` flag](https://developer.apple.com/documentation/foundation/nsurlisexcludedfrombackupkey).

On Android I moved the installation ID from `SharedPreferences` of name `host.exp.exponent.SharedPreferences` (shared across ExpoKit, so I wouldn't exclude it from the backup too lightly) to `SharedPreferences` of a separate `expo.modules.notifications.InstallationId`. I then added `fullBackupContent` configuration which automatically configures exclusion of this file unless other library does the same or developer overrides the `android:fullBackupContent` configuration. For this situation I have added a README section explaining how to consolidate configurations of different libraries into one.

# Test Plan

Debugging natively I have ensured that migration from legacy to new storage works as expected. I wasn't able to test whether the new storage is not backed up (`adb shell bmgr backupnow` didn't work as expected and I'm not up for backing up and restoring my iOS device, I did however verify that the `expo-notifications-installation-id.txt` file has `com.apple.metadata:com_apple_backup_excludeItem` attribute using instructions from [here](https://www.howtobuildsoftware.com/index.php/how-do/R71/ios-cordova-icloud-how-to-check-which-files-are-stored-on-icloud)).

### CODE REVIEW

**Comment by sjchmiela** [comment]
Adding @ide as a reviewer as I remember you reviewing this feature while it was developed in the first place – you may have opinions on this too.

A maybe interesting side note — if the server accepted not [only `uuidv4`](https://github.com/expo/universe/blob/d40f6005f5ed96129b65aaa62f6156e83763b544/server/www/src/apiv2/PushRouter.ts#L40) as `deviceId`, but other strings too we could use Firebase `InstanceId` as recommended in [Android documentation](https://developer.android.com/training/articles/user-data-ids#best-practices-android-identifiers) and not worry about Android backups at all (and a somewhat difficult setup instructions).

**Comment by ide** [comment]
@sjchmiela The server needs a UUID but it doesn't need to be UUID v4. We can use UUID v5 with a hardcoded "namespace" and the Android Instance ID as the "name". UUID v5 typically gives 122 bits of randomness, which is comparable to UUID v4.

I actually think this Instance ID approach is better than trying to exclude the files from backup so that we don't have to deal with a file that could become corrupt, have fsync issues, etc...

Is there a similar approach we could take on iOS? (I understand it doesn't use most of the code in this PR but I think it will be simpler to maintain.)

File: packages/expo-notifications/README.md [hunk 1]
```diff
@@ -47,6 +47,44 @@ In order to be able to receive push notifications on the device ensure that your
 
 This module requires permission to subscribe to device boot. It's used to setup the scheduled notifications right after the device (re)starts. The `RECEIVE_BOOT_COMPLETED` permission is added automatically.
 
+`expo-notifications` automatically configures your Android app to omit internal installation ID data storage (required for Expo push token to work properly) from device backups. However, if you or any other library customize [`android:fullBackupContent` property of `<application>`](https://developer.android.com/guide/topics/manifest/application-element#fullBackupContent) you will need to consolidate all the rules manually:
```
Comment (cruzach) [inline packages/expo-notifications/README.md:50]:
```suggestion
`expo-notifications` automatically configures your Android app to omit internal installation ID data storage (required for the Expo push token to work properly) from device backups. However, if you or any other library customize the [`android:fullBackupContent` property of `<application>`](https://developer.android.com/guide/topics/manifest/application-element#fullBackupContent), you will need to add all the rules manually:
```

File: packages/expo-notifications/README.md [hunk 2]
```diff
@@ -47,6 +47,44 @@ In order to be able to receive push notifications on the device ensure that your
...
+- create an XML file in (usually) `android/app/src/main/res/xml/` directory, name it eg. `backup_descriptor.xml`
+- go to `AndroidManifest.xml` of your app, find `<application>` tag and add the `android:fullBackupContent` attribute of value `@xml/name_of_file_you_just_created_without_extension`. If you named your file `backup_descriptor.xml`, your `<application>` tag should look like this:
```
Comment (cruzach) [inline packages/expo-notifications/README.md:52]:
```suggestion
- go to your `AndroidManifest.xml`, find `<application>` tag, and add the `android:fullBackupContent` attribute of value `@xml/name_of_file_you_just_created_without_extension`. If you named your file `backup_descriptor.xml`, your `<application>` tag should look like this:
```

File: packages/expo-notifications/README.md [hunk 3]
```diff
@@ -47,6 +47,44 @@ In order to be able to receive push notifications on the device ensure that your
...
+    android:fullBackupContent="@xml/backup_descriptor"
+    ...>
+  ```
+- go back to the XML file you just referenced and inside it write:
+  ```xml
+  <?xml version="1.0" encoding="utf-8"?>
+  <full-backup-content>
+  </full-backup-content>
+  ```
+  this makes a good starting point for more configurations.
```
Comment (cruzach) [inline packages/expo-notifications/README.md:65]:
```suggestion
  This is a good starting point for more configurations.
```

File: packages/expo-notifications/README.md [hunk 4]
```diff
@@ -47,6 +47,44 @@ In order to be able to receive push notifications on the device ensure that your
...
+- a section required by `expo-notifications` is:
```
Comment (cruzach) [inline packages/expo-notifications/README.md:66]:
```suggestion
- **The following is required by `expo-notifications`**:
```

File: packages/expo-notifications/README.md [hunk 5]
```diff
@@ -47,6 +47,44 @@ In order to be able to receive push notifications on the device ensure that your
...
+  ```xml
+  <exclude
+    domain="sharedpref"
+    path="expo.modules.notifications.InstallationId.xml" />
+  ```
+  add it in between `<full-backup-content>` and `</full-backup-content>`.
+- go through other libraries and add their configurations as described in their instructions.
```
Comment (cruzach) [inline packages/expo-notifications/README.md:73]:
```suggestion
- Go through the rest of your libraries and add their configurations as described in their instructions.
```

File: packages/expo-notifications/README.md [hunk 6]
```diff
@@ -47,6 +47,44 @@ In order to be able to receive push notifications on the device ensure that your
...
+- in the end the file should look more or less like this
```
Comment (cruzach) [inline packages/expo-notifications/README.md:74]:
```suggestion
- In the end the file should look more or less like this:
```

File: packages/expo-notifications/README.md [hunk 7]
```diff
@@ -47,6 +47,44 @@ In order to be able to receive push notifications on the device ensure that your
...
+  <full-backup-content>
+    <!-- expo-notifications -->
+    <exclude
+      domain="sharedpref"
+      path="expo.modules.notifications.InstallationId.xml" />
+    
+    <!-- ... other configurations -->
+  </full-backup-content>
+  ```
+- the process has also been described in Android documentation, so if you want to read more head over to [_Include and exclude files_ section of _Back up user data with Auto Backup_ doc](https://developer.android.com/guide/topics/data/autobackup#IncludingFiles).
```
Comment (cruzach) [inline packages/expo-notifications/README.md:86]:
```suggestion
- This process has also been described in Android's documentation, so if you want to read more head over to [_Include and exclude files_ section of the _Back up user data with Auto Backup_ doc](https://developer.android.com/guide/topics/data/autobackup#IncludingFiles).
```

File: packages/expo-notifications/ios/EXNotifications/EXInstallationIdProvider.m [hunk 8]
```diff
@@ -15,12 +17,81 @@ @implementation EXInstallationIdProvider
...
+      // If there was no error when persisting legacy installation ID in new storage
+      // we are safe to clear the legacy storage.
+      [self removeLegacyInstallationId];
+    }
+    return legacyInstallationId;
+  }
+  
+  // There's no installation ID available, neither legacy nor current one.
+  installationId = [self generateNewInstallationId];
+  NSError *error = [self persistInstallationId:legacyInstallationId];
```
Comment (cruzach) [inline packages/expo-notifications/ios/EXNotifications/EXInstallationIdProvider.m:45]: should this be:
```suggestion
  NSError *error = [self persistInstallationId:installationId];
```

**Review by cruzach — CHANGES_REQUESTED** [review]
mostly wording changes to readme, plus one change in `EXInstallationIdProvider.m`

**Comment by ide** [comment]
On iOS, there's [`identifierForVendor`](https://developer.apple.com/documentation/uikit/uidevice/1620059-identifierforvendor). I believe we expose this in `expo-device` as well. This field alone is not enough because two apps from the same developer will share an identifier. But we could perhaps do something like `uuidv5(namespace = hardcoded value, name = UIDevice.current.identifierForVendor + bundle identifier)`.

We need to be careful with this field as it is not reliable in all scenarios:

> If the value is nil, wait and get the value again later. This happens, for example, after the device has been restarted but before the user has unlocked the device.
>
> The value in this property remains the same while the app (or another app from the same vendor) is installed on the iOS device. The value changes when the user deletes all of that vendor’s apps from the device and subsequently reinstalls one or more of them. The value can also change when installing test builds using Xcode or when installing an app on a device using ad-hoc distribution. Therefore, if your app stores the value of this property anywhere, you should gracefully handle situations where the identifier changes.

Maybe there is a different field that is better. Or we could use the non-backed-up file.

**Comment by sjchmiela** [comment]
Thanks for help, James, I came to similar conclusions!

---

For Android that could actually work quiet well—we would move the responsibility for making sure it's unique per device to Firebase and get rid of complicated installation instructions, and given Firebase' Instance ID has to be available by the time we try to register Expo push token (for device push token to be available, an instance ID has to be assigned), this sounds like a good approach.

---

When it comes to iOS I found another interesting way—we could use [`kSecAttrAccessibleAlwaysThisDeviceOnly `](https://developer.apple.com/documentation/security/ksecattraccessiblealwaysthisdeviceonly?language=objc) and store the installation ID in Keychain. This has the advantage of not being so user-visible (as is a file in a documents directory) and the disadvantage of being deprecated (even though still usable, it looks like it's not going anywhere, like the deprecation is just a strong _do not use in applications, because why would you_). We could also try to use its non-deprecated variation `WhenUnlocked` or `AfterFirstUnlock` (resulting in behavior equal or better than `identifierForVendor + bundleId`).

---

~Yet another idea would be to use a [`DeviceCheck API`](https://developer.apple.com/documentation/devicecheck?language=objc) which seems like it does what we need:~

> ~Use the shared instance of the `DCDevice` class to generate a token that identifies a device.~

~Unfortunately it's supported only since iOS 11 and requires an Internet connection.~ Unfortunately,

> You should treat the token you receive in the completion block as single-use. Although the token remains valid long enough for your server to retry a specific request if necessary, you should not use a token multiple times. Instead, use this method to generate a new token.

---

We could also depend on Firebase on iOS too just for the instance ID although I'm not a fan of this solution.

**Comment by ide** [comment]
I don't think we should use a deprecated field. It's just going to cause more work for us if it is removed and there is no simple way to retain the same behavior. `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly` looks interesting, and we could just throw an error with a clear error code if the field is unavailable. Are there cases where a device has restarted and is still locked, and would want to register for notifications? (I suppose an app could receive a notification and run background code that re-registers for notifications. In this case we'd just skip re-registration if it's our own native code doing it, or throw a JS error if an app's code is registering for notifications.)

One thing we need to be very careful about with keychain is to write code that is "self-healing" if we ever get into a bad state (e.g. if an invalid device ID gets stored) because uninstalling and reinstalling an app does not clear the keychain. (It was supposed to clear data after Apple saw that Uber was permanently storing state using the keychain but then Apple rolled back that behavior.)

Pros of identifierForVendor: we don't store state ourselves
Cons of identifierForVendor: if Apple changes this API or algorithm we will need to migrate away from it to something like the keychain. identifierForVendor's behavior seems less well-defined for development apps. We also can't use identifierForVendor for existing installations since they already have their own device ID.

Pros of keychain: we need a place to store the device ID for all existing installations anyway -- just one code path to think about after a device's ID has been migrated.
Cons of keychain: persists across uninstalling and reinstalling app so mistakes are especially expensive (but I trust you @sjchmiela). We will also need to think of a way to store the info that will not collide with other modules that access the keychain.

DeviceCheck is for reading the 2 bits of permanent info. I don't think it helps us here.

> We could also depend on Firebase on iOS too just for the instance ID although I'm not a fan of this solution.

Strongly agree that depending on a proprietary library will be costly in many ways.


Also as for Constants.installationID - we have two options:

- Migrate it to the keychain as well so that we only have to think about one ID
- Keep it separate so that we preserve the existing semantics

Either way I would like to deprecate this field in favor of developers just generating a random ID and storing it themselves. There's no need for Expo to provide this.

**Comment by ide** [comment]
Thinking about this some more, I think it'd also be an improvement to make this change just for Android and think about iOS later, since it's more clear what to do on Android.

**Comment by sjchmiela** [comment]
One problem I think we have is that it will be very difficult or impossible to implement this in a future-proof way without sacrificing all the device IDs rotating.

The problem is that the easy way to keep old ID if it's defined and use new ID otherwise is that virtually we'll keep using old IDs everywhere — the logic would go like
```
if (existing installation ID is present) {
  use existing
}

use new firebase-based ID
```
which neither helps with device backups that will be created from now on (existing ID will remain there and new devices restored from backups will still have IDs of older devices) nor prevents conflicts on new installations (there are more places where the UUID is created, namely: `ExponentNotificationIntentService` which will need to remain in `expoview` until we remove legacy `Notifications` and `expo-constants`' binding in `expoview` which exposes (getting-or-creating) the ID as module constant.

I think in order to not make all tokens rotate we'd need to write a manual migration that would move existing installation ID from backuped storage to non-backed up storage where we can write values by hand (so using Firebase Installation ID doesn't really cut it)—excluded-from-backup `SharedPrefences` files and keychain | excluded-from-backup text file.

**Comment by ide** [comment]
As we discussed:

- We need to migrate the current device ID to a location that is excluded from backups.
- We are going to apply this change to both `Constants.installationId` and the device ID for the Expo push notification service. (Both use cases will continue to use the exact same ID.)
- `Constants.installationId` will eventually be deprecated since it can be implemented in userspace, so we should optimize for push notifications instead. Other libraries probably shouldn't try to use this ID, so putting it in code that is clearly related to notifications so others don't try to use it seems helpful.

Proposal:
On both Android and iOS, we will need to store the installation ID for existing apps in a location that is excluded from backups. On Android, this is a separated SharedPreferences instance that is excluded from backups (is there a better way to exclude data from backups?). On iOS, this is a keychain entry with `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly`.

We run the migration lazily, only when the installation ID is requested. The logic would look like:

1. Check if the new SharedPreferences/keychain entry exists.
2. If we didn't read any value, we potentially need to migrate the ID from the old synced SharedPreferences/NSUserDefaults.
  1. Check if there is an ID in the old location.
  2. If it exists, write the ID to the new SharedPreferences/keychain entry.
  3. Delete the old SharedPreferences/NSUserDefaults entry. Migration complete.
3. At this point, we either have an ID from the new SharedPreferences/keychain, an ID from the old SharedPreferences/NSUserDefaults (which has been moved to the new location), or no ID at all. The old SharedPreferences/keychain entry is gone.
4. If we have an ID, check if it is a valid UUID. If it's valid, return it. Otherwise, delete the SharedPreferences/keychain entry so we recover from corrupt data.
5. If we didn't read a valid ID (because there was no ID or because it was corrupt and we deleted it), we need to generate a new ID. This is the code path that all new installations will take.
6. On Android, we compute the v5 UUID from the Firebase Instance ID. We don't store any information in SharedPreferences -- it's a pure function (assuming Firebase Instance ID is pure).
7. On iOS, we generate a random v4 UUID and write it to the keychain. That is, IDs for new installations are treated the same as IDs for old installations that have been migrated. The reason we don't try to write a pure function on iOS is because `identiferForVendor` isn't as precisely documented as the keychain API, specifically around when it returns nil.
8. If we wanted Android and iOS to behave more similarly, we could ignore the Firebase Instance ID and just store a v4 UUID in SharedPreferences. (I lean towards using Firebase because the configuration to exclude SharedPreferences from backup sounds breakable. In contrast, the keychain API on iOS has better documentation and clearer behavior.)

---

To address the issue of `ExponentNotificationIntentService` in expoview: what if we had a small, shared class that's used by both Notifications and Constants?

**Comment by sjchmiela** [comment]
Superseded by https://github.com/expo/expo/pull/11005 and https://github.com/expo/expo/pull/11019.

**Pull request closed without merging**